File: WINGS_AI_Bootcamp/Python/NumToWords.py
# Number to words converter - American and Indian numbering systems

ONES = ["", "One", "Two", "Three", "Four", "Five", "Six", "Seven", "Eight", "Nine"]
TEENS = ["Ten", "Eleven", "Twelve", "Thirteen", "Fourteen", "Fifteen",
         "Sixteen", "Seventeen", "Eighteen", "Nineteen"]
TENS = ["", "", "Twenty", "Thirty", "Forty", "Fifty", "Sixty", "Seventy", "Eighty", "Ninety"]

def two_digit(num):
    if num == 0:
        return ""
    elif num < 10:
        return ONES[num]
    elif 10 <= num < 20:
        return TEENS[num - 10]
    else:
        return TENS[num // 10] + (" " + ONES[num % 10] if num % 10 != 0 else "")

def three_digit(num):
    hundred = num // 100
    rest = num % 100
    res = ""
    if hundred > 0:
        res += ONES[hundred] + " Hundred"
        if rest > 0:
            res += " "
    if rest > 0:
        res += two_digit(rest)
    return res

# American system scales
AMERICAN_SCALE = ["", "Thousand", "Million", "Billion", "Trillion"]

def num_to_words_american(num):
    if num == 0:
        return "Zero"
    words = []
    i = 0
    while num > 0:
        part = num % 1000
        if part > 0:
            segment = three_digit(part)
            if AMERICAN_SCALE[i]:
                segment += " " + AMERICAN_SCALE[i]
            words.append(segment.strip())
        num //= 1000
        i += 1
    return " ".join(words[::-1]).strip()

# Indian system scales
INDIAN_SCALE = ["", "Thousand", "Lakh", "Crore", "Arab"]

def num_to_words_indian(num):
    if num == 0:
        return "Zero"
    words = []
    # first 3 digits
    part = num % 1000
    words.append(three_digit(part))
    num //= 1000
    i = 1
    while num > 0 and i < len(INDIAN_SCALE):
        part = num % 100
        if part > 0:
            segment = two_digit(part)
            if INDIAN_SCALE[i]:
                segment += " " + INDIAN_SCALE[i]
            words.append(segment.strip())
        num //= 100
        i += 1
    return " ".join(filter(None, words[::-1])).strip()

# Testing

print("American System:")
print(num_to_words_american(123456789))
print(num_to_words_american(1000100))
print(num_to_words_american(0))

print("\nIndian System:")
print(num_to_words_indian(123456789))
print(num_to_words_indian(1000100))
print(num_to_words_indian(0))

File: WINGS_AI_Bootcamp/Haskell/NumToWords.hs
import Data.List (intercalate)

space, andWord, hundred :: String
space = " "
andWord = "and"
hundred = "hundred"

upto20 :: [String]
upto20 = ["", "One", "Two", "Three", "Four", "Five", "Six", "Seven", "Eight", "Nine",
          "Ten", "Eleven", "Twelve", "Thirteen", "Fourteen", "Fifteen",
          "Sixteen", "Seventeen", "Eighteen", "Nineteen"]

tens :: [String]
tens = ["", "", "Twenty", "Thirty", "Forty", "Fifty", "Sixty", "Seventy", "Eighty", "Ninety"]

convert2 :: Int -> String
convert2 k
  | k < 20 = upto20 !! k
  | otherwise = tens !! (k `div` 10) ++ (if k `mod` 10 /= 0 then space ++ upto20 !! (k `mod` 10) else "")

convert3 :: Int -> String
convert3 k
  | k < 100 = convert2 k
  | r == 0 = convert2 h ++ space ++ hundred
  | otherwise = convert2 h ++ space ++ hundred ++ space ++ andWord ++ space ++ convert2 r
  where (h, r) = k `divMod` 100

type System = [(Int, String)]

western, indian :: System
western = [(10^12, "Trillion"), (10^9, "Billion"), (10^6, "Million"), (1000, "Thousand"), (1, "")]
indian = [(10^9, "Arab"), (10^7, "Crore"), (10^5, "Lakh"), (1000, "Thousand"), (1, "")]

toWords :: System -> Int -> String
toWords system amount = intercalate space . filter (/= "") $ go system amount
  where
    go [] _ = []
    go ((denom, label):rest) amt =
      let (q, r) = amt `divMod` denom
      in (if q > 0 then convert3 q ++ (if label /= "" then space ++ label else "") else "") : go rest r

main :: IO ()
main = do
  putStrLn $ toWords western 671
  putStrLn $ toWords indian 123456789
